### ISSUE TITLE
RequestSpecificRetryHandler for RibbonLoadBalancingHttpClient does not have the correct connection related exceptions

### ISSUE THREAD

**Issue by jebeaudet** [spring-cloud/spring-cloud-netflix#1439]
In the class `RibbonLoadBalancingHttpClient` (which now extends `AbstractLoadBalancingClient` in Camden), the `RequestSpecificRetryHandler` is a generic one and does not apply to HttpClient. 

Specifically, the exceptions that are connection related are only exceptions extending `SocketException`. This is not the case for HttpClient so POST/PUT/DELETE won't get retried properly since they don't have the flag `okToRetryOnAllErrors`.

Ribbon has an implementation for HttpClient : `com.netflix.niws.client.http.HttpClientLoadBalancerErrorHandler` that list the connect related exceptions, I think we should use those while keeping the `RequestSpecificRetryHandler` implementation. I'd even add `java.net.ssl.SSLHandshakeException` since this can be thrown if the server shuts down in the middle of an handshake.

**Comment by jebeaudet** [comment]
We have the same problem with `FeignLoadBalancer` actually, if you use httpclient with it you won't get proper retry on connect exceptions

**Comment by ryanjbaxter** [comment]
@jebeaudet this retry logic is going to change.

Here are my changes for Feign
https://github.com/spring-cloud/spring-cloud-netflix/pull/1457

Changes for Zuul will be next.

**Comment by jebeaudet** [comment]
I'm looking forward to see it in action! However, from the quick look I had doesn't that just wraps ribbon into another retry logic layer? Feign already had this with the `Retryer` interface actually. I think

**Comment by jebeaudet** [comment]
I'm looking forward to see it in action! However, from the quick look I had doesn't that "just" wraps ribbon into another retry logic layer? I find it cumbersome to have multiple layers of retrying (like in the feign ribbon client, you have the ribbon retry logic and the `Retryer` interface on top of it), it's hard to follow, hard to configure and hard to debug.

To be honest, I'm not a fan of Ribbon, the Observable pattern is impossible to follow and debug and it's overly complicated IMO. I'd be quite happy to ditch it entirely in favor of spring retry and a more unified code path for different clients.

What do you think? I don't want to sound overly criticizing, just trying to make this lib better :)

Thanks

**Comment by ryanjbaxter** [comment]
It is not adding another layer.  We are disabling both the [ribbon retry logic](https://github.com/spring-cloud/spring-cloud-netflix/pull/1457/files#diff-6a0e8e400f382e1041b27d7e06aaa39cR124) and [feign retry logic](https://github.com/spring-cloud/spring-cloud-netflix/pull/1457/files#diff-177f70f71e8d0a77b90c2b31b4255ec4R109).

> To be honest, I'm not a fan of Ribbon, the Observable pattern is impossible to follow and debug and it's overly complicated IMO. I'd be quite happy to ditch it entirely in favor of spring retry and a more unified code path for different clients.

Agree this is why we are doing this :)

Feel free to try out my branch and see it in action.  Would love your feedback.  Can we close this issue and work through the existing issue?

**Comment by jebeaudet** [comment]
OK I see, the retry logic is still present for `ClientException` though, maybe a custom implementation that always return false for the `isRetriableException()` method would be cleaner than passing false everywhere I'd say. 

I'll try to check out your branch and run it through my test suite. Thanks I'll close this issue for now.
